pcdsdevices/analog_signals.py

```diff
@@ -61,6 +61,56 @@ class Acromag(BaseInterface, Device):
     tab_component_names = True
 
 
+class AcromagChannelInput(BaseInterface, Device):
+    """
+    Class for an Acromag Channel Input.
+
+    Parameters
+    ----------
+    prefix : str
+        The base EPICS PV for the Acromag. E.g.: `XPP:USR:ai1`
+    channel : str
+        The channel number. E.g.: `1` [0-15]
+    name : str
+        A name to refer to the Acromag Input Channel.
+    """
+    channel = FCpt(EpicsSignalRO, '{self._prefix}:{self._channel}',
```

Comment:
I'm wondering about the possibility of the class inheriting from EpicsSignal instead of having one as a component. I guess that would make the class a `Signal`, not a `Device`. Does that hurt it?

I'm also still not fully set on these classes being necessary at all, since they really only contain a single Signal.

If that's not the right path, could we instead have methods for setting and getting the voltage on the device level so we can do `ao1.put()` instead of `ao1.channel.put()`. It looks like a small change but I can already hear the scientists unhappy about having to type the extra word.

Reply:
I am not sure here how I would do it with inheriting from EpicsSignal... but I guess an alternative is to have a factory method like you guys mentioned before, instead of having two devices here, but this will be a bit more tricky as I need to differentiate from an Input/Output channel, and also I need to create a device with only One channel instead of all of them..... my understanding was that the scientist won't use these two devices outside the questionnaire? 

Need to think about this more though... thank you for the suggestions 😊

Reply:
A factory function would be the alternative here. We briefly had chatted about the options here and I seem to recall we landed on "Device", but maybe I'm misremembering.

I like that we can use `FCpt` here - that's the nicest part of the Device option.

Reply:
Yeah I thought we landed on 'Device' too... but I also thought that I could have misinterpreted it hahah, so it's not bad to revise this and re-think about the other options pros&cons 😊 ...
I'll leave it as is for now then. 
Also, I can do a `get()` on the `a` instead of `a.channel.get()` - but the output does not look as pretty, i'm also assuming i can do an `a.put()` but won't try that. - 
```
In [2]: a = AcromagChannelInput(prefix='XPP:USR:ai1', channel='1')

In [3]: a
Out[3]:

ai
--
channel: -0.008544921875 [Volts]

In [4]: a.get()
Out[4]: AcromagChannelInputTuple(channel=-0.0079345703125)

In [5]: a.channel.get()
Out[5]: -0.008544921875

```
@ZryletTC did you mean that I should have something like this:??
```
def get_voltage(self):
    self.channel.get()

def set_voltage(self, voltage):
    self.channel.put(voltage)
```

Reply:
Not sure where best to comment: 
In the general beamline python, I will likely have an object for my acromag with my AI card and its channels and similar for the AI and DI/DO,... cards. I use this to e.g. define LaserShuttters of other devices that take AO channels in input which are permanently installed. For experiment setup that changes (where requests are put in through the questionnaire), I have two things happening: for an AI channels, I might need to read the value in some function that would write for a given experiment (i.e. wait until signal reaches XXX level, then runs the sequencer). For an AO channel, I might want to write to that AO to e.g start a device, close a shutter,... in some script combined with other stuff or just by itself. 
Currently, entering channels in he questions creates a while acromag object (or at least the whole card) which is definitely not what I want. Maybe all that should happen is to link/alias the corresponding channel on my beamline acromag to the name that is defined in the questionnaire, so that e.g. 'LED.put(5)' will set xpp.acromag.ao1.channelXXX to 5 V (note: I obviously have not looked up what exactly the my acromag is called)

Reply:
Ok maybe i need to do a bit more digging here because i'm not sure anymore if i understood correctly the requirements for this issue. 
So before when loading the devices form the experiment we'd get something like this: 
```
 'acromag':
acromag
-------
ao1_0: None
ao1_1: None
ao1_2: None
ao1_3: None
ao1_4: None
ao1_5: None
ao1_6: None
ao1_7: None
ao1_8: None
ao1_9: None
ao1_10: None
ao1_11: None
ao1_12: None
ao1_13: None
ao1_14: None
ao1_15: None
ai1_0: None
ai1_1: None
ai1_2: None
ai1_3: None
ai1_4: None
ai1_5: None
ai1_6: None
ai1_7: None
ai1_8: None
ai1_9: None
ai1_10: None
ai1_11: None
ai1_12: None
ai1_13: None
ai1_14: None
ai1_15: None}
```
and now we're getting something like this:
```
 'ao1_6':
ao_6
----
channel: 0.0 [Volts]}
```

Where from the questionnaire run details we had valid info only for `ao_1` - 
```
 'pcdssetup-ao-6-device': '',
 'pcdssetup-ao-6-name': '',
 'pcdssetup-ao-6-pvbase': '',
 'pcdssetup-ao-1-pvbase': 'XPP:USR:ao1',
 'pcdssetup-ao-4-purpose': '',
 'pcdssetup-ao-5-purpose': '',
 'pcdssetup-ao-5-device': '',
 'pcdssetup-ao-4-device': '',
 'pcdssetup-ao-5-name': '',
 'pcdssetup-ao-4-name': '',
 'pcdssetup-ao-4-pvbase': '',
 'pcdssetup-ao-5-pvbase': '',
 'pcdssetup-ao-1-purpose': 'Evo Shutter 1',
 'pcdssetup-ao-1-channel': '6',
 'pcdssetup-ao-2-channel': '7',
 'pcdssetup-ao-3-channel': '9',
 'pcdssetup-ao-3-purpose': '',
 'pcdssetup-ao-2-purpose': '',
 'pcdssetup-ao-2-device': '',
 'pcdssetup-ao-3-device': '',
 'pcdssetup-ao-2-name': '',
 'pcdssetup-ao-3-name': '',
 'pcdssetup-ao-3-pvbase': '',
 'pcdssetup-ao-2-pvbase': '',
 'pcdssetup-ao-1-name': 'acromag',
```
Am I on the wrong path here?

Reply:
@silkenelson I believe is saying she doesn't want the entire Acromag device, but the individual channels as you are working on. So, I think you're overall on the right path here, @cristinasewell.

As to whether or not those individual signals are borrowed from the full `Acromag` device or are created one-by-one (through either a factory function or Device): well, these are implementation details where we have some wiggle room as long as we deliver what the scientists want.

Reply:
Hey Cristina, yeah, you're definitely on the right path, and the yes those functions will be good if you just add a return.

As for the other implementation details, I was just throwing out ideas. I think it'd be cool if we could have a single acromag device and make the questionnaire borrow signals from it but if that's too much work, this way works just as well.

One note if we do leave it this way with channel devices, we might want to use those inside the Acromag class, instead of the plain signals.

Reply:
Thanks @ZryletTC. 
One thing here is that the way I create the Acromag Channels from the questionnaire differs a bit from the actual Acromag class. 
For the Acromag Channels from the questionnaire I have the prefix as: `XPP:USR:ao1` and then I just add the channel... something like this: `AcromagChannelOutput(prefix='XPP:USR:ao1', channel='6')`.
And currently when we create an Acromag device we would do something like this: `Acromag(prefix='XPP:USR')` - so different `prefix` expected.
Using these guys inside the `Acromag` class might require some changes to either the way we create an Acromag device or Acromag Channels - I don't think we want to change the way we create the Acromag device, correct?

Reply:
You're right that we don't want to change the way we create Acromag devices, but the way an Acromag defines its channel components should be a pretty simple change. I believe it should just be the following:
from
```ao1_0 = Cpt(EpicsSignal, ":ao1:0", kind='normal')```
to
```ao1_0 = Cpt(AcromagChannelOutput, ":ao1", channel=0, kind='normal')```

Reply:
Ohh I see :D - thank you

Reply:
One downside to note is that if we do this, the API for the class will be slightly different. That can be curtailed by overriding the get and put methods but still kinda eh 😕 
@klauer or @ZLLentz, what do you think?

Reply:
Given that this thread hasn't closed yet and we're still seemingly stuck on this being a `Device` - my take is that a factory function is really what's desired here, such that we get a simple `EpicsSignal` out of it instead of a `Device`. I personally think that's just fine.

Reply:
Ok, you guys know better, I'll go with whatever you think is best. 

So, should I have a factory function that will basically determine if it is an Input or Output base on the `prefix` - `XPP:USR:ai1` vs `XPP:USR:ao1`?

Reply:
How about use the prefix-based assumption, but allow it to be overridden by an argument?
Perhaps

```python
def factory_func(prefix, *, is_input=None, ...):
    if is_input is None:
        is_input = ':ai' in prefix
```

or

```python
def factory_func(prefix, *, signal_class=None, ...):
    if signal_class is None:
        signal_class = EpicsSignalRO if ':ai' in prefix else EpicsSignal
```

Reply:
ok sounds good, thank you @klauer